`TypeError` and empty filenames when using re.Pattern functionality of MultiZarrToZarr’s `coo_mapper`

I’ve ran into and addressed two issues while using the regex pattern functionality of `coo_mapper` on `MultiZarrToZarr`. It’s unclear whether these reflect intentional design decisions so I’m opening an issue to clarify whether a PR implementing my fixes is warranted or not.

These problems were encountered using kerchunk version `0.0.6` on Python `3.8.4`.

**Problem statement**
I am preprocessing a collection of NetCDFs that were originally standalone rasters (BILs) and thus lack a `time` dimension. Fortunately, their dates are expressed as 8-digit strings in the filename, i.e. `file_20210915.nc`. Therefore I generate the time dimension using a Regex pattern with the `coo_map` setting on `MultiZarrToZarr`:

```
nc_list = glob.glob(input_dir, “*.nc4”)
json_list = []
for nc in nc_list:
            fs = fsspec.filesystem('file')
            fs_nc = fs.open(nc)
            with fs_nc as infile:
                json_list.append(SingleHdf5ToZarr(infile, fs_nc.path).translate())

pattern = re.compile(r'([0-9]{8})') # extract 8 digit YYYYYmmdd strings from file names
mzz_opts = dict(    identical_dims = [“lat’, “lon”],
                    coo_map = {"time" : pattern},
                    concat_dims = [“time”])
mzz = MultiZarrToZarr(json_list, **mzz_opts)
mzz.translate()
```
This returns the following error
```
    def _get_value(self, index, z, var, fn=None):
        """derive coordinate value(s) for given input dataset

        How to map from input to

        index: int
            Current place in the list of inputs
        z: zarr group
            Open for the current input
        var: str
            name of value to extract.
        fn: str
            filename
        """
        selector = self.coo_map[var]
        if isinstance(selector, collections.abc.Callable):
            o = selector(index, z, var, fn)
        elif isinstance(selector, list):
            o = selector[index]
        elif isinstance(selector, re.Pattern):
>           o = selector.match(fn).groups[0]
E           TypeError: expected string or bytes-like object
```
The problem appears to be the use of `re.match` instead of `re.search` — match always starts from the 0 index of an object whereas my datetime strings are at the end of variable length strings. I can resolve this error by amending `combine.py` line 154 from `o = selector.match(fn).groups[0]` to `o = re.search(selector,fn)[0]`.

More problematically, I found that `combine.py`’s `fss` property generates an empty list of filenames. Replacing line 116 (`self._paths = [None] * len(fo_list)` as below resolves the issue:

```
if isinstance(self.path[0], collections.abc.Mapping):
    self._paths = []
    fo_list = self.path
    # self._paths = [None] * len(fo_list) # replaced line
    for path in self.path: # new line
        self._paths.append(path["templates"]["u"]) # new line
```

The problem appears to be kerchunk dict objects being recognized as `collections.abc.Mapping` instances. Reviewing the code, it’s unclear to me if this is intentional or not. If the latter I’m happy to open a PR.

A small collection of sample files used for this operation can be found at https://www.dropbox.com/sh/yzbqqkqampcjcqo/AACKPVnB9bojkF5QVRZihU3za?dl=0.

--- a/kerchunk/combine.py
+++ b/kerchunk/combine.py
@@ -155,7 +155,7 @@ class MultiZarrToZarr:
                 fo_list = self.path
                 self._paths = []
                 for path in self.path:
-                    self._paths.append(path["templates"]["u"])
+                    self._paths.append(path.get("templates", {}).get("u", None))
             else:
                 self._paths = []
                 for of in fsspec.open_files(self.path, **self.target_options):
@@ -195,7 +195,7 @@ class MultiZarrToZarr:
         elif isinstance(selector, list):
             o = selector[index]
         elif isinstance(selector, re.Pattern):
-            o = re.search(selector,fn)[0]
+            o = selector.search(fn)[0]
         elif not isinstance(selector, str):
             # constant, should be int or float
             o = selector

--- a/kerchunk/combine.py
+++ b/kerchunk/combine.py
@@ -153,7 +153,9 @@ class MultiZarrToZarr:
                 self._paths = self.path
             elif isinstance(self.path[0], collections.abc.Mapping):
                 fo_list = self.path
-                self._paths = [None] * len(fo_list)
+                self._paths = []
+                for path in self.path:
+                    self._paths.append(path["templates"]["u"])
             else:
                 self._paths = []
                 for of in fsspec.open_files(self.path, **self.target_options):
@@ -193,7 +195,7 @@ class MultiZarrToZarr:
         elif isinstance(selector, list):
             o = selector[index]
         elif isinstance(selector, re.Pattern):
-            o = selector.match(fn).groups()[0]  # may raise
+            o = re.search(selector,fn)[0]
         elif not isinstance(selector, str):
             # constant, should be int or float
             o = selector

--- a/kerchunk/combine.py
+++ b/kerchunk/combine.py
@@ -195,7 +195,7 @@ class MultiZarrToZarr:
         elif isinstance(selector, list):
             o = selector[index]
         elif isinstance(selector, re.Pattern):
-            o = selector.search(fn)[0]
+            o = selector.search(fn).groups()[0]
         elif not isinstance(selector, str):
             # constant, should be int or float
             o = selector

